Mathlib/Tactic/Simproc/Divisors.lean

```diff
@@ -0,0 +1,63 @@
+/-
+Copyright (c) 2025 Paul Lezeau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Paul Lezeau, Bhavik Mehta
+-/
+import Mathlib.NumberTheory.Divisors
+import Mathlib.Util.Qq
+import Mathlib.Data.Finset.Sort
+
+/-! # Divisor Simprocs
+
+This file implements simprocs to compute various objects related to divisors:
+- `Nat.divisorsEq`: computes `Nat.divisors n` for explicit values of `n`
+- `Nat.properDivisorsEq`: computes `Nat.properDivisors n` for explicit values of `n`
+
+-/
+
+open Lean Meta Qq
+
+/-- The `Nat.divisorsEq` computes the finset `Nat.divisors n` when `n` is a natural number
+litteral. -/
+simproc_decl Nat.divisorsEq (Nat.divisors _) := fun e => do
+  unless e.isAppOfArity `Nat.divisors 1 do return .continue
+  let some n ← fromExpr? e.appArg! | return .continue
+  let rhsListQ : List Q(ℕ) := (n.divisors.sort (· ≤ ·)).map fun n => (Lean.toExpr n : Q(ℕ))
+  let rhs := mkSetLiteralQ q(Finset ℕ) rhsListQ
+  /- The last two lines can be replace by `return .done {expr := rhs }` since the proof is `rfl`.
+  Not sure what's best here. -/
+  let pf ← Meta.mkDecideProof (← mkEq e rhs)
+  return .done {expr := rhs, proof? := pf }
+
+/-- The `Nat.properDivisorsEq ` computes the finset `Nat.properDivisorsEq  n` when `n` is a natural
+number litteral. -/
+simproc_decl Nat.properDivisorsEq (Nat.properDivisors _) := fun e => do
+  unless e.isAppOfArity `Nat.properDivisors 1 do return .continue
+  let some n ← fromExpr? e.appArg! | return .continue
+  let rhsListQ : List Q(ℕ) := (n.properDivisors.sort (· ≤ ·)).map fun n => (Lean.toExpr n : Q(ℕ))
+  let rhs := mkSetLiteralQ q(Finset ℕ) rhsListQ
+  /- The last two lines can be replace by `return .done {expr := rhs }` since the proof is `rfl`.
+  Not sure what's best here. -/
+  let pf ← Meta.mkDecideProof (← mkEq e rhs)
+  return .done {expr := rhs, proof? := pf }
+
+example : Nat.divisors 1710 = {1, 2, 3, 5, 6, 9, 10, 15, 18, 19, 30, 38, 45, 57,
+      90, 95, 114, 171, 190, 285, 342, 570, 855, 1710} := by
+  simp only [Nat.divisorsEq]
+
+example : Nat.divisors 57 = {1, 3, 19, 57} := by
+  simp only [Nat.divisorsEq]
+
+example : 2 ≤ Finset.card (Nat.divisors 3) := by
```

Comment:
Let's test what happens if the input to `Nat.divisors` is a non-numeral (in part because I can't see which part of the current code is checking that)

Reply:
Added one. The piece of the code checking for this is `fromExpr? e.appArg!`, which fails when `n` isn't a litteral.